### PR TITLE
BUG: Avoid Timedelta rounding when specifying unit and integer (#12690)

### LINKED ISSUE #12690: Rounding errors when parsing timedeltas
Found this while working on #12136.

Rounding errors occur for timedeltas initialized using fractional microsecond values:

```
>>> pd.Timedelta('12.312us') == pd.Timedelta(12312)
False
>>> pd.Timedelta(12.312, unit='us') == pd.Timedelta(12312)
False
>>> pd.Timedelta('12.312us') == pd.Timedelta(12311)
True
```

Additionally, precision is lost when initializing timedeltas with fractional millisecond (and probably second) values:

```
>>> pd.Timedelta(0.012312, unit='ms') == pd.Timedelta(12311)
False
>>> pd.Timedelta(0.012311, unit='ms') == pd.Timedelta(12311)
False
>>> pd.Timedelta(0.012311, unit='ms') == pd.Timedelta(12000)
True
```

I believe the problem is here: https://github.com/pydata/pandas/blob/288059adfb751d52be57e444198198518d256c05/pandas/tslib.pyx#L3399
bumping up the values of `p` should probably solve this.

#### output of ``pd.show_versions()``

commit: None
python: 2.7.11.final.0
python-bits: 64
OS: Windows
OS-release: 10
machine: AMD64
processor: Intel64 Family 6 Model 78 Stepping 3, GenuineIntel
byteorder: little
LC_ALL: None
LANG: None

pandas: 0+unknown
nose: 1.3.7
pip: 8.1.0
setuptools: 20.2.2
Cython: 0.23.4
numpy: 1.10.4
scipy: 0.17.0
statsmodels: None
xarray: None
IPython: 4.1.2
sphinx: 1.3.5
patsy: 0.4.0
dateutil: 2.4.2
pytz: 2015.7
blosc: None
bottleneck: 1.0.0
tables: 3.2.2
numexpr: 2.4.6
matplotlib: 1.5.1
openpyxl: 2.3.2
xlrd: 0.9.4
xlwt: 1.0.0
xlsxwriter: 0.8.4
lxml: 3.5.0
bs4: 4.3.2
html5lib: 0.999
httplib2: None
apiclient: None
sqlalchemy: 1.0.12
pymysql: 0.6.7.None
psycopg2: None
jinja2: 2.8
boto: None

### LINKED ISSUE #14156: BUG: conversion precision on Timedeltas
I think this is rounding prematurely, though not entirely sure. Could also just be loss of precision when converting from a float -> int64.

```
In [12]: pd.to_datetime(1434743731.8770001,unit='s')
Out[12]: Timestamp('2015-06-19 19:55:31.877000')

In [13]: pd.to_datetime(1434743731.8770001,unit='s').value
Out[13]: 1434743731877000000

In [14]: (pd.Timestamp('2015-06-19 19:55:15.877000')+pd.Timedelta('1us')).value
Out[14]: 1434743715877001000

```

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] closes #14156
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

File: pandas/tests/scalar/test_timedelta.py [hunk 1]
```diff
@@ -432,6 +432,16 @@ def test_construction(self):
...
 
+    @pytest.mark.parametrize("unit, value, expected", [
+        ('us', 9.999, 9999), ('ms', 9.999999, 9999999),
+        ('s', 9.999999999, 9999999999)])
+    def test_rounding_on_int_unit_construction(self, unit, value, expected):
+        # GH 12690
+        result = Timedelta(value, unit=unit)
+        assert result.value == expected
+        result = Timedelta(str(value) + unit)
+        assert result.value
```
Comment (jschendel) [inline pandas/tests/scalar/test_timedelta.py]: looks like the last line is missing `== expected`

File: doc/source/whatsnew/v0.23.0.txt [hunk 2]
```diff
@@ -716,6 +716,7 @@ Datetimelike
 - Bug in :class:`Timestamp` and :func:`to_datetime` where a string representing a barely out-of-bounds timestamp would be incorrectly rounded down instead of raising ``OutOfBoundsDatetime`` (:issue:`19382`)
 - Bug in :func:`Timestamp.floor` :func:`DatetimeIndex.floor` where time stamps far in the future and past were not rounded correctly (:issue:`19206`)
 - Bug in :func:`to_datetime` where passing an out-of-bounds datetime with ``errors='coerce'`` and ``utc=True`` would raise ``OutOfBoundsDatetime`` instead of parsing to ``NaT`` (:issue:`19612`)
+- Bug in :class:`Timedelta`: where a numerical value with a unit would round values (:issue: `12690`)
```
Comment (jschendel) [inline doc/source/whatsnew/v0.23.0.txt]: there's an extra colon at the end of :class:\`Timedelta\`: that should be removed, and the space between :issue: and the number should also be removed

**Comment by mroeschke** [comment]
Although I believe these rounding constants are correct, it appears there are precision issues now for larger integers.

```
# Example from the failing test

In [6]: v
Out[6]: 946688461.0005

In [7]: pd._libs.tslibs.timedeltas.cast_from_unit(v, 's')
Out[7]: 946688461000499968
```

File: pandas/_libs/tslibs/timedeltas.pyx [hunk 3]
```diff
@@ -200,22 +200,22 @@ cpdef inline int64_t cast_from_unit(object ts, object unit) except? -1:
 
     if unit == 'D' or unit == 'd':
         m = 1000000000L * 86400
-        p = 6
```
Comment (jreback) [inline pandas/_libs/tslibs/timedeltas.pyx]: these were set originally to avoid precision issues.

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_libs/tslibs/timedeltas.pyx [hunk 4]
```diff
@@ -229,10 +229,10 @@ cpdef inline int64_t cast_from_unit(object ts, object unit) except? -1:
     # cast the unit, multiply base/frace separately
     # to avoid precision issues from float -> int
     base = <int64_t> ts
-    frac = ts -base
+    frac = ts - base
     if p:
         frac = round(frac, p)
```
Comment (jreback) [inline pandas/_libs/tslibs/timedeltas.pyx]: you might be able to do something like this

``round(frac * 1000, p - 3) // 1000`` (for p >- 3) but haven't looked really closely.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by mroeschke** [comment]
Unfortunately using a 1000 multiplier didn't solve the issue, @jreback.

The issue with the example above is with floating point errors before rounding.

```
In [2]: v = 946688461.0005

In [3]: v - int(v)
Out[3]: 0.0004999637603759766

In [4]: round(v - int(v), 6) # precision in master
Out[4]: 0.0005

In [5]: round(v - int(v), 9) # precision in this PR (in order to preserve nanoseconds if specified)
Out[5]: 0.000499964
```

Would it be reasonable to cast `v` to a string in order to parse the whole number from the decimal?

```
In [8]: str(v).split('.')
Out[8]: ['946688461', '0005']
```

**Comment by jreback** [comment]
you cannot use strings ; rather you might look how CPython does this

**Comment by jorisvandenbossche** [comment]
To come back to this, the question is if we should care. In the example of `v = 946688461.0005`, even in the full number (not only after splitting the fractional part), the actual precision is 946688461.00049996. So this is inherent problem of floating points even in the input, not only in our calculation.

As a user you can always solve this by eg doing `*1000` your data and specify a different `unit`.

**Comment by Sup3rGeo** [comment]
Also complementing @jorisvandenbossche, I also believe that the current rounding behavior is way, way more detrimental to users than this precision issue with the changes. 

This means that even if we care about this precision issue, it would be better to merge this changes (reflecting correctly the timedelta value for a given integer and unit) and just open another issue on the precision thing for large numbers.

**Comment by mroeschke** [comment]
I agree as well that premature rounding is less desirable than floating point errors. I could issue a warning for now for floats with high significant digits.

**Comment by jorisvandenbossche** [comment]
>  I could issue a warning for now for floats with high significant digits.

Could we easily detect the case when the number of decimals is too high to preserve precision? In that case, a warning might be a good idea (then we can also give a hint on how to solve it)

**Comment by mroeschke** [comment]
@jorisvandenbossche after searching I don't think there's a reliable way to detect significant digits since not all floats are stored exactly.

An alternative for now is to live with the precision issue and offer a note in the docs?

**Comment by jorisvandenbossche** [comment]
> An alternative for now is to live with the precision issue and offer a note in the docs?

Yes, that would be my preference. @jreback what's your view on this?

**Comment by jreback** [comment]
sure ok with simply documenting this as a known limitation. (doc-string and/or main docs)

**Comment by mroeschke** [comment]
Sorry for the delay on this PR.

I had to adjust some existing tests to account for floating point errors after this fix. Additionally, it appears we already have a note in the main docs about floating point errors when creating timestamps at the bottom of this section:

https://pandas.pydata.org/pandas-docs/stable/timeseries.html#epoch-timestamps

Ready for a final look and all green.

File: pandas/tests/scalar/timestamp/test_timestamp.py [hunk 5]
```diff
@@ -623,7 +623,7 @@ def test_basics_nanos(self):
 
     def test_unit(self):
 
-        def check(val, unit=None, h=1, s=1, us=0):
```
Comment (jreback) [inline pandas/tests/scalar/timestamp/test_timestamp.py]: would be nice to parameterize this test

File: pandas/tests/io/sas/test_sas7bdat.py [hunk 6]
```diff
@@ -182,6 +182,11 @@ def test_date_time():
     fname = os.path.join(dirpath, "datetime.csv")
     df0 = pd.read_csv(fname, parse_dates=['Date1', 'Date2', 'DateTime',
                                           'DateTimeHi', 'Taiw'])
+    # GH 19732: Timestamps imported from sas will incur floating point errors
```
Comment (jreback) [inline pandas/tests/io/sas/test_sas7bdat.py]: why did this need to be changed?

**Review by jreback — CHANGES_REQUESTED** [review]
can you show a copy-pastable example which works now (and didn't before). I know your tests cover this, but  easier to copy-paste it.

**Comment by jreback** [comment]
@mroeschke to my last, is there an explicit test from the OP?

File: pandas/tests/io/sas/test_sas7bdat.py (hunk 6 above)
Comment (mroeschke) [reply pandas/tests/io/sas/test_sas7bdat.py]: I believe there are floating point errors when importing dates from the sas file.

```
In [5]: pd.read_sas('pandas/tests/io/sas/data/datetime.sas7bdat')['DateTimeHi']
Out[5]:
0   1677-09-21 00:12:43.145225525
1   1960-01-01 00:00:00.000000000
2   2016-02-29 23:59:59.123456001
3   2262-04-11 23:47:16.854774475
Name: DateTimeHi, dtype: datetime64[ns]

# This matches dates in datetime.csv
In [6]: pd.read_csv('pandas/tests/io/sas/data/datetime.csv')['DateTimeHi']
Out[6]:
0    1677-09-21 00:12:43.145226
1    1960-01-01 00:00:00.000000
2    2016-02-29 23:59:59.123456
3    2262-04-11 23:47:16.854774
Name: DateTimeHi, dtype: object
```

I am not familiar with read_sas or the sas file that was created, but I am fairly certain it's due to the floating point errors

**Comment by mroeschke** [comment]
Added an explicit test from the original issue (which was from you @jreback) and paramed an existing test.

File: pandas/tests/io/sas/test_sas7bdat.py (hunk 6 above)
Comment (jreback) [reply pandas/tests/io/sas/test_sas7bdat.py]: can you round these instead of specifying exact values

File: doc/source/whatsnew/v0.23.1.txt [hunk 7]
```diff
@@ -45,7 +45,11 @@ Bug Fixes
 ~~~~~~~~~
 
 -
--
+
+Timedelta
+^^^^^^^^^
+- Bug in :class:`Timedelta`: where a numerical value with a unit would round values (:issue: `12690`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.1.txt]: can you be a little bit precise here, e.g. i think a reader might not get what this change is

**Review by jreback — CHANGES_REQUESTED** [review]
small comments, otherwise lgtm.

File: doc/source/whatsnew/v0.23.1.txt [hunk 8]
```diff
@@ -53,7 +53,10 @@ Strings
 ^^^^^^^
 
 - Bug in :meth:`Series.str.replace()` where the method throws `TypeError` on Python 3.5.2 (:issue: `21078`)
--
+
+Timedelta
+^^^^^^^^^
+- Bug in :class:`Timedelta`: where passing a float with a unit would prematurely round the float precision (:issue: `12690`)
```
Comment (jreback) [inline doc/source/whatsnew/v0.23.1.txt]: shouldn't this be 14156?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by mroeschke** [comment]
Thanks @jreback. I had reference the older issue that was replaced by OPs.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @mroeschke

**Comment by jorisvandenbossche** [comment]
As this is strictly spoken an API change for those that used a float with less precision and now get higher precision with floating point errors, I would rather keep this for 0.24.0 ?

**Pull request merged**